foundations/src/telemetry/memory_profiler.rs

```diff
@@ -182,26 +189,67 @@ fn init_profiler(settings: &MemoryProfilerSettings) -> BootstrapResult<Option<Me
     Ok(Some(MemoryProfiler {
         _seal: Seal,
 
-        #[cfg(feature = "security")]
-        sandbox_profiling_syscalls: settings.sandbox_profiling_syscalls,
+        request_heap_profile: request_sender,
+        _heap_profiling_thread_handle: Arc::new(Mutex::new(heap_profile_thread_handle)),
     }))
 }
 
-fn collect_heap_profile() -> Result<String> {
+#[cfg(feature = "security")]
+fn receive_profiling_thread_setup_msg(
+    setup_complete_receiver: mpsc::Receiver<anyhow::Result<()>>,
+) -> anyhow::Result<()> {
+    use std::sync::mpsc::RecvTimeoutError;
+
+    const PROFILING_THREAD_SETUP_TIMEOUT: Duration = Duration::from_secs(2);
+    match setup_complete_receiver.recv_timeout(PROFILING_THREAD_SETUP_TIMEOUT) {
+        Ok(Ok(())) => {}
+        Ok(Err(setup_err)) => {
+            return Err(setup_err);
+        }
+        Err(RecvTimeoutError::Timeout) => bail!("Profiling thread did not finish setup in time"),
+        Err(RecvTimeoutError::Disconnected) => {
+            bail!("Profiling thread disconnected before finishing setup")
+        }
+    }
+
+    Ok(())
+}
+
+fn heap_profile_thread(
+    receive_request: mpsc::Receiver<oneshot::Sender<anyhow::Result<String>>>,
+    #[cfg(feature = "security")] setup_complete: mpsc::Sender<anyhow::Result<()>>,
+    sandbox_profiling_syscalls: bool,
+) {
+    #[cfg(feature = "security")]
+    if sandbox_profiling_syscalls {
+        if let Err(_) = setup_complete.send(sandbox_jemalloc_syscalls()) {
+            return;
```

Comment:
Should we log something here? I didn't see much of a point as this can only happen when the main/parent thread drops the `Receiver` corresponding to this `Sender`, which should only happen in error scenarios, where something else is already going wrong.

Same question for the loop below

Reply:
Yeah if the main thread is not there to receive from this channel, then I agree something else must have already happened that would be observed.

Reply:
Worth dropping a comment in this function to explain that though

Reply:
Added comment